include/fc/io/cfile.hpp

```diff
@@ -65,77 +80,118 @@ class cfile {
+
+       size_t result = fread( d, 1, n, _file.get() );
+
+       if( result != n ) {
+           throw std::ios_base::failure( "cfile: " + _file_path.generic_string() +
                                        " unable to read " + std::to_string( n ) + " bytes; only read " + std::to_string( result ) );
-      }
+       }
    }
 
    void write( const char* d, size_t n ) {
-      size_t result = fwrite( d, 1, n, _file.get() );
-      if( result != n ) {
-         throw std::ios_base::failure( "cfile: " + _file_path.generic_string() +
+       if(!_open)
+           throw std::ios_base::failure("cfile is not open");
+
+       size_t result = fwrite( d, 1, n, _file.get() );
+
+       if( result != n ) {
+           throw std::ios_base::failure( "cfile: " + _file_path.generic_string() +
                                        " unable to write " + std::to_string( n ) + " bytes; only wrote " + std::to_string( result ) );
-      }
+       }
    }
 
    void flush() {
-      if( 0 != fflush( _file.get() ) ) {
-         int ec = ferror( _file.get() );
-         throw std::ios_base::failure( "cfile: " + _file_path.generic_string() +
+       if(!_open)
+           throw std::ios_base::failure("cfile is not open");
+
+       if( 0 != fflush( _file.get() ) ) {
+           int ec = ferror( _file.get() );
+           throw std::ios_base::failure( "cfile: " + _file_path.generic_string() +
                                        " unable to flush file, ferror: " + std::to_string( ec ) );
-      }
+       }
    }
 
    void sync() {
-      const int fd = fileno(_file.get() );
-      if( -1 == fd ) {
-         throw std::ios_base::failure( "cfile: " + _file_path.generic_string() +
+       if(!_open)
+           throw std::ios_base::failure("cfile is not open");
+
+       const int fd = fileno(_file.get() );
+
+       if( -1 == fd ) {
+           throw std::ios_base::failure( "cfile: " + _file_path.generic_string() +
                                        " unable to convert file pointer to file descriptor, error: " +
                                        std::to_string( errno ) );
-      }
-      if( -1 == fsync( fd ) ) {
-         throw std::ios_base::failure( "cfile: " + _file_path.generic_string() +
+       }
+
+       if( -1 == fsync( fd ) ) {
+           throw std::ios_base::failure( "cfile: " + _file_path.generic_string() +
                                        " unable to sync file, error: " + std::to_string( errno ) );
-      }
+       }
    }
 
-   bool eof() const { return feof(_file.get()) != 0; }
+    bool eof() const  {
+        if(!_open)
+            throw std::ios_base::failure("cfile is not open");
 
-   int getc() { 
-      int ret = fgetc(_file.get());  
-      if (ret == EOF) {
-         throw std::ios_base::failure( "cfile: " + _file_path.generic_string() +
+        if(feof(_file.get()))return true;
+
+        int c = fgetc(_file.get());
+        if(EOF == c)return true;
+
+        return (EOF == ungetc(c, _file.get()));
```

Comment:
Is this complicated version fo `eof()` needed? Can you provide some information about why these changes are needed. I'm a bit concerned about the `ungetc`:
"A call to fseek, fsetpos or rewind on stream will discard any characters previously put back into it with this function."

Reply:
i was being paranoid. I removed it.